File: app/src/main/java/by/anegin/telegram_contests/core/ui/view/MiniChartView.java
package by.anegin.telegram_contests.core.ui.view;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.view.ViewParent;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.Interpolator;
import by.anegin.telegram_contests.R;
import by.anegin.telegram_contests.core.ui.ScaleAnimationHelper;
import by.anegin.telegram_contests.core.ui.model.Graph;
import by.anegin.telegram_contests.core.ui.model.UiChart;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MiniChartView extends View implements ScaleAnimationHelper.Callback {

    private static final int DRAG_RANGE = 1;
    private static final int DRAG_RANGE_START = 2;
    private static final int DRAG_RANGE_END = 3;
    private static final int DRAG_NONE = 4;

    private static final int TOGGLE_ANIMATION_DURATION = 200;

    private final Paint windowPaint = new Paint();
    private final Paint touchRipplePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint fadePaint = new Paint();

    private float windowStrokeWidthTopBottom;
    private float windowStrokeWidthLeftRight;

    private float chartLineWidth;

    private float rangeStart = 0f;
    private float rangeEnd = 1f;

    private float downX = 0f;
    private float lastTouchX = 0f;
    private boolean inDragMode = false;

    private int dragMode = DRAG_NONE;

    private int touchSlop;
    private float dragSize;

    private Animator touchRippleAnimation;
    private float touchRippleRadius = 0f;
    private float touchRippleAlpha = 0f;

    private ChartView attachedChartView;

    private final ScaleAnimationHelper scaleAnimationHelper = new ScaleAnimationHelper(this, TOGGLE_ANIMATION_DURATION);

    private UiChart uiChart;
    private List<Graph> graphs;

    private final Map<String, ValueAnimator> hideAnimators = new HashMap<>();
    private final Map<String, ValueAnimator> showAnimators = new HashMap<>();

    private final Matrix transformMatrix = new Matrix();
    private float xOffs;
    private float yOffs;
    private float xScale;
    private float yScale;

    public MiniChartView(Context context) {
        super(context);
        init(context, null, 0);
    }

    public MiniChartView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs, 0);
    }

    public MiniChartView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    private void init(Context context, AttributeSet attrs, int defStyleAttr) {
        setSaveEnabled(true);

        TypedArray viewAttrs = context.obtainStyledAttributes(attrs, R.styleable.MiniChartView, defStyleAttr, 0);
        rangeStart = viewAttrs.getFloat(R.styleable.MiniChartView_range_start, 0f);
        rangeEnd = viewAttrs.getFloat(R.styleable.MiniChartView_range_end, 1f);
        int fadeColor = viewAttrs.getColor(R.styleable.MiniChartView_fade_color, Color.TRANSPARENT);
        int windowColor = viewAttrs.getColor(R.styleable.MiniChartView_window_color, Color.TRANSPARENT);
        windowStrokeWidthTopBottom =
                viewAttrs.getDimension(R.styleable.MiniChartView_window_stroke_width_top_bottom, 0f);
        windowStrokeWidthLeftRight =
                viewAttrs.getDimension(R.styleable.MiniChartView_window_stroke_width_left_right, 0f);
        chartLineWidth = viewAttrs.getDimension(R.styleable.MiniChartView_chart_line_width, 1f);
        viewAttrs.recycle();

        if (rangeStart < 0f) rangeStart = 0f;
        if (rangeEnd > 1f) rangeEnd = 1f;

        windowPaint.setStyle(Paint.Style.FILL);
        windowPaint.setColor(windowColor);

        touchRipplePaint.setStyle(Paint.Style.FILL);
        touchRipplePaint.setColor(windowColor);

        fadePaint.setStyle(Paint.Style.FILL);
        fadePaint.setColor(fadeColor);

        touchSlop = ViewConfiguration.get(context).getScaledTouchSlop();
        dragSize = 1.5f * touchSlop;
    }

    @Override
    public void onDraw(Canvas canvas) {
        int width = getWidth();
        int height = getHeight();
        if (width == 0 || height == 0) return;

        float rangeStartX = width * rangeStart;
        float rangeEndX = width * rangeEnd;

        // draw window
        float x1 = rangeStartX + windowStrokeWidthLeftRight / 2f;
        float x2 = rangeEndX - windowStrokeWidthLeftRight / 2f;
        windowPaint.setStrokeWidth(windowStrokeWidthLeftRight);
        canvas.drawLine(x1, 0f, x1, height, windowPaint);
        canvas.drawLine(x2, 0f, x2, height, windowPaint);
        float x3 = rangeStartX + windowStrokeWidthLeftRight;
        float x4 = rangeEndX - windowStrokeWidthLeftRight;
        float y1 = windowStrokeWidthTopBottom / 2f;
        float y2 = height - windowStrokeWidthTopBottom / 2f;
        windowPaint.setStrokeWidth(windowStrokeWidthTopBottom);
        canvas.drawLine(x3, y1, x4, y1, windowPaint);
        canvas.drawLine(x3, y2, x4, y2, windowPaint);

        // draw Chart data
        List<Graph> graphs = this.graphs;
        if (graphs != null) {
            canvas.save();
            canvas.clipRect(0f, 0f, width, height);
            for (Graph graph : graphs) {
                if (graph.state != Graph.STATE_HIDDEN) {
                    graph.draw(canvas);
                }
            }
            canvas.restore();
        }

        // fade out off-window regions
        if (rangeStartX > 0f) {
            canvas.drawRect(0f, 0f, rangeStartX, height, fadePaint);
        }
        if (rangeEndX < width) {
            canvas.drawRect(rangeEndX, 0f, width, height, fadePaint);
        }

        // draw touch ripple
        float touchRippleRadius = this.touchRippleRadius;
        if (touchRippleRadius > 0f) {
            Float cx = null;
            switch (dragMode) {
                case DRAG_RANGE_START: {
                    cx = rangeStartX + windowStrokeWidthLeftRight / 2f;
                    break;
                }
                case DRAG_RANGE_END: {
                    cx = rangeEndX - windowStrokeWidthLeftRight / 2f;
                    break;
                }
                case DRAG_RANGE: {
                    cx = rangeStartX + (rangeEndX - rangeStartX) / 2f;
                    break;
                }
            }
            if (cx != null) {
                touchRipplePaint.setAlpha((int) (102 * touchRippleAlpha));
                canvas.drawCircle(cx, height / 2f, touchRippleRadius, touchRipplePaint);
            }
        }
    }

    @Override
    @SuppressLint("ClickableViewAccessibility")
    public boolean onTouchEvent(MotionEvent event) {
        int action = event.getAction();
        switch (action) {
            case MotionEvent.ACTION_DOWN:
                float rangeStartX = getWidth() * rangeStart;
                float rangeEndX = getWidth() * rangeEnd;

                downX = event.getX();
                lastTouchX = downX;
                inDragMode = false;

                // determine drag mode depending on touch region
                if (downX >= rangeStartX - dragSize && downX <= rangeStartX + dragSize + windowStrokeWidthLeftRight) {
                    dragMode = DRAG_RANGE_START;
                } else if (downX >= rangeEndX - dragSize - windowStrokeWidthLeftRight && downX <= rangeEndX + dragSize) {
                    dragMode = DRAG_RANGE_END;
                } else if (downX >= rangeStartX + dragSize + windowStrokeWidthLeftRight && downX <= rangeEndX - dragSize - windowStrokeWidthLeftRight) {
                    dragMode = DRAG_RANGE;
                } else {
                    dragMode = DRAG_NONE;
                }

                showTouchRippleAnimation();
                break;
            case MotionEvent.ACTION_MOVE:
                float touchX = event.getX();
                if (!inDragMode && Math.abs(touchX - downX) > touchSlop) {
                    inDragMode = true;

                    ViewParent viewParent = getParent();
                    if (viewParent != null) {
                        viewParent.requestDisallowInterceptTouchEvent(true);
                    }
                }
                if (inDragMode && dragMode != DRAG_NONE) {
                    onMove(touchX, lastTouchX - touchX);
                }
                lastTouchX = touchX;
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL: {
                hideTouchRippleAnimation();
                break;
            }
        }
        return true;
    }

    private void onMove(float eventX, float distanceX) {
        int width = getWidth();
        float rangeStartX = width * rangeStart;
        float rangeEndX = width * rangeEnd;

        switch (dragMode) {
            case DRAG_RANGE_START: {
                // move left edge of range window (0..rangeEnd)
                float newRangeStart = eventX / width;
                if (newRangeStart < 0f) newRangeStart = 0f;

                float newRangeStartX = width * newRangeStart;
                if (newRangeStartX + dragSize + windowStrokeWidthLeftRight > rangeEndX - dragSize - windowStrokeWidthLeftRight) {
                    newRangeStartX = rangeEndX - 2f * (dragSize + windowStrokeWidthLeftRight);
                    newRangeStart = newRangeStartX / width;
                }
                updateRanges(newRangeStart, rangeEnd, true);
                break;
            }
            case DRAG_RANGE_END: {
                // move right edge of range window (rangeStart..1)
                float newRangeEnd = eventX / width;
                if (newRangeEnd > 1f) newRangeEnd = 1f;

                float newRangeEndX = width * newRangeEnd;
                if (newRangeEndX - dragSize - windowStrokeWidthLeftRight < rangeStartX + dragSize + windowStrokeWidthLeftRight) {
                    newRangeEndX = rangeStartX + 2f * (dragSize + windowStrokeWidthLeftRight);
                    newRangeEnd = newRangeEndX / width;
                }
                updateRanges(rangeStart, newRangeEnd, true);
                break;
            }
            case DRAG_RANGE: {
                // move whole window
                float newRangeStart = (rangeStartX - distanceX) / width;
                if (newRangeStart < 0f) newRangeStart = 0f;

                float newRangeEnd = newRangeStart + (rangeEnd - rangeStart);
                if (newRangeEnd > 1f) {
                    newRangeEnd = 1f;
                    newRangeStart = 1f - (rangeEnd - rangeStart);
                }
                updateRanges(newRangeStart, newRangeEnd, true);
                break;
            }
        }
    }

    private void updateRanges(float start, float end, boolean notifyListener) {
        if (rangeStart != start || rangeEnd != end) {
            rangeStart = start;
            rangeEnd = end;

            invalidate();

            if (notifyListener && attachedChartView != null) {
                attachedChartView.setRange(start, end, true);
            }
        }
    }

    private void showTouchRippleAnimation() {
        if (touchRippleAnimation != null) {
            touchRippleAnimation.cancel();
        }
        touchRippleAnimation = createTouchRippleAnimation(0f, 1f);
        touchRippleAnimation.start();
    }

    private void hideTouchRippleAnimation() {
        float currentValue = touchRippleAnimation != null ? (float) ((ValueAnimator) touchRippleAnimation).getAnimatedValue() : 1f;
        if (touchRippleAnimation != null) {
            touchRippleAnimation.cancel();
        }
        touchRippleAnimation = createTouchRippleAnimation(currentValue, 0f);
        touchRippleAnimation.start();
    }

    private Animator createTouchRippleAnimation(float from, float to) {
        ValueAnimator anim = ValueAnimator.ofFloat(from, to);
        anim.setDuration(200);
        anim.setInterpolator(new DecelerateInterpolator());
        anim.addUpdateListener(animation -> {
            float value = (float) animation.getAnimatedValue();
            touchRippleRadius = value * (0.7f * getHeight());
            touchRippleAlpha = value;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                postInvalidateOnAnimation();
            } else {
                postInvalidate();
            }
        });
        return anim;
    }

    // ================

    @Override
    public void onSizeChanged(int w, int h, int oldw, int oldh) {
        enqueueUpdateData();
    }

    public void attachToChartView(ChartView chartView) {
        this.attachedChartView = chartView;

        chartView.setOnUiChartChangeListener(uiChart -> {
            this.uiChart = uiChart;
            enqueueUpdateData();
        });

        chartView.setOnRangeChangeListener((start, end) -> updateRanges(start, end, false));

        chartView.setOnGraphVisibilityChangeListener(new ChartView.OnGraphVisibilityChangeListener() {
            @Override
            public void onGraphHidden(String id) {
                hideGraph(id);
            }

            @Override
            public void onGraphShown(String id) {
                showGraph(id);
            }

            @Override
            public void onReset() {
                showAllGraphs();
            }
        });

        chartView.setRange(rangeStart, rangeEnd, false);
    }

    private void enqueueUpdateData() {
        UiChart uiChart = this.uiChart;
        if (uiChart == null) return;

        float viewWidth = getWidth();
        xScale = 1f * viewWidth / uiChart.width;
        xOffs = (viewWidth - uiChart.width * xScale) / 2f;
        yOffs = getHeight() * 0.925f;

        List<Graph> newGraphs = new ArrayList<>();
        for (Graph g : uiChart.graphs) {
            newGraphs.add(new Graph(g, chartLineWidth));
        }

        this.graphs = newGraphs;

        scaleAnimationHelper.calculate(false);
    }

    // ===============

    private void hideGraph(String id) {
        Graph graph = findGraph(id);
        if (graph == null) return;

        if (graph.state != Graph.STATE_HIDING) {
            graph.state = Graph.STATE_HIDING;
            scaleAnimationHelper.calculate(true);
        }

        // cancel show animation if exists
        ValueAnimator showAnimation = showAnimators.remove(id);
        if (showAnimation != null && showAnimation.isRunning()) {
            showAnimation.cancel();
        }

        // add hide animation if not exists
        if (!hideAnimators.containsKey(id)) {
            ValueAnimator hideAnimation = makeToggleAnimation(graph, 0f,
                    Graph.STATE_HIDING, Graph.STATE_HIDDEN,
                    new DecelerateInterpolator(),
                    () -> hideAnimators.remove(id));
            hideAnimators.put(id, hideAnimation);
            hideAnimation.start();
        }
    }

    private void showGraph(String id) {
        Graph graph = findGraph(id);
        if (graph == null) return;

        if (graph.state != Graph.STATE_SHOWING) {
            graph.state = Graph.STATE_SHOWING;
            scaleAnimationHelper.calculate(true);
        }

        // cancel hide animation if exists
        ValueAnimator hideAnimation = hideAnimators.remove(id);
        if (hideAnimation != null && hideAnimation.isRunning()) {
            hideAnimation.cancel();
        }

        // add show animation if not exists
        if (!showAnimators.containsKey(id)) {
            ValueAnimator showAnimation = makeToggleAnimation(graph, 1f,
                    Graph.STATE_SHOWING, Graph.STATE_VISIBLE,
                    new AccelerateInterpolator(),
                    () -> showAnimators.remove(id));
            showAnimators.put(id, showAnimation);
            showAnimation.start();
        }
    }

    private Graph findGraph(String id) {
        for (Graph g : this.graphs)
            if (g.id.equals(id)) return g;
        return null;
    }

    private ValueAnimator makeToggleAnimation(Graph graph, float endValue, int progressState, int finalState, Interpolator interpolator, Runnable onEndAction) {
        float startValue = graph.animationValue;
        long duration = (long) (TOGGLE_ANIMATION_DURATION * Math.abs(endValue - startValue));

        ValueAnimator anim = ValueAnimator.ofFloat(startValue, endValue);
        anim.setInterpolator(interpolator);
        anim.setDuration(duration);
        anim.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                onEndAction.run();
                if (graph.state == progressState) {
                    graph.state = finalState;
//                    graph.animationValue = endValue;
                    invalidate();
                }
            }
        });
        anim.addUpdateListener(animation -> {
            if (graph.state == progressState) {
                float value = (float) animation.getAnimatedValue();
                graph.animationValue = value;
                Log.v("ABC", "value: " + value);
                invalidate();
            } else {
                animation.cancel();
            }
        });
        return anim;
    }

    private void showAllGraphs() {
        List<Graph> graphs = this.graphs;
        for (Graph graph : graphs) {
            graph.state = Graph.STATE_VISIBLE;
            graph.animationValue = 1f;
            scaleAnimationHelper.calculate(false);
        }

        for (ValueAnimator anim : hideAnimators.values()) {
            if (anim.isRunning()) anim.cancel();
        }
        hideAnimators.clear();
        for (ValueAnimator anim : showAnimators.values()) {
            if (anim.isRunning()) anim.cancel();
        }
        showAnimators.clear();

        invalidate();
    }

    // ===============

    @Override
    public float calculateNewScale() {
        float maxHeight = 0f;
        for (Graph graph : graphs) {
            float max = graph.findMaxYInRange(0f, uiChart.width);
            if (max > maxHeight) maxHeight = max;
        }
        return 0.85f * getHeight() / maxHeight;
    }

    @Override
    public float getCurrentScale() {
        return yScale;
    }

    @Override
    public void onScaleUpdated(float scale) {
        yScale = scale;

        transformMatrix.reset();
        transformMatrix.setTranslate(xOffs, yOffs);
        transformMatrix.preScale(xScale, -yScale);

        for (Graph graph : this.graphs) {
            graph.transform(transformMatrix);
        }

        postInvalidate();
    }

    // ===============

    @Override
    protected Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();
        SavedState savedState = new SavedState(superState);
        savedState.rangeStart = rangeStart;
        savedState.rangeEnd = rangeEnd;
        return savedState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        SavedState savedState = (SavedState) state;
        super.onRestoreInstanceState(savedState.getSuperState());
        rangeStart = savedState.rangeStart;
        rangeEnd = savedState.rangeEnd;
    }

    private static class SavedState extends BaseSavedState {

        float rangeStart;
        float rangeEnd;

        SavedState(Parcelable superState) {
            super(superState);
        }

        private SavedState(Parcel in) {
            super(in);
            rangeStart = in.readFloat();
            rangeEnd = in.readFloat();
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeFloat(rangeStart);
            out.writeFloat(rangeEnd);
        }

        public static final Parcelable.Creator<SavedState> CREATOR = new Parcelable.Creator<SavedState>() {
            public SavedState createFromParcel(Parcel in) {
                return new SavedState(in);
            }

            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }

}

File: app/src/main/java/by/anegin/telegram_contests/core/ui/view/MiniChartView.kt
package by.anegin.telegram_contests.core.ui.view

import android.animation.Animator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.DecelerateInterpolator
import by.anegin.telegram_contests.R
import by.anegin.telegram_contests.core.ui.model.Graph
import by.anegin.telegram_contests.core.ui.model.UiChart
import java.util.concurrent.Executors

class MiniChartView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    companion object {
        private const val DRAG_RANGE = 1
        private const val DRAG_RANGE_START = 2
        private const val DRAG_RANGE_END = 3
        private const val DRAG_NONE = 4
    }

    private val windowPaint = Paint()
    private val touchRipplePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val chartPaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG)
    private val fadePaint = Paint()

    private val windowStrokeWidthTopBottom: Float
    private val windowStrokeWidthLeftRight: Float

    private var rangeStart = 0f
    private var rangeEnd = 1f

    private var downX = 0f
    private var lastTouchX = 0f
    private var inDragMode = false

    private var dragMode: Int? = null

    private val touchSlop: Int
    private val dragSize: Float

    private var touchRippleAnimation: Animator? = null
    private var touchRippleRadius = 0f
    private var touchRippleAlpha = 0f

    private var onRangeChangeListener: ((Float, Float) -> Unit)? = null

    private val updateDataExecutor = Executors.newFixedThreadPool(2)
    private var lastUpdateGeneration = 0L

    private var uiChart: UiChart? = null
    private var graphs: List<Graph>? = null

    init {
        val typedValue = TypedValue()
        val defAttrs = context.obtainStyledAttributes(typedValue.data, intArrayOf(android.R.attr.colorPrimary))
        val primaryColor = defAttrs.getColor(0, 0)
        defAttrs.recycle()

        val viewAttrs = context.obtainStyledAttributes(attrs, R.styleable.MiniChartView, defStyleAttr, 0)
        rangeStart = viewAttrs.getFloat(R.styleable.MiniChartView_range_start, 0f)
        rangeEnd = viewAttrs.getFloat(R.styleable.MiniChartView_range_end, 1f)
        val offWindowAlpha = viewAttrs.getFloat(R.styleable.MiniChartView_off_window_alpha, 0.5f)
        val windowBgColor = viewAttrs.getColor(R.styleable.MiniChartView_window_bg_color, primaryColor)
        windowStrokeWidthTopBottom =
            viewAttrs.getDimension(R.styleable.MiniChartView_window_stroke_width_top_bottom, 0f)
        windowStrokeWidthLeftRight =
            viewAttrs.getDimension(R.styleable.MiniChartView_window_stroke_width_left_right, 0f)
        val chartLineWidth = viewAttrs.getDimension(R.styleable.MiniChartView_chart_line_width, 1f)
        viewAttrs.recycle()

        if (rangeStart < 0f) rangeStart = 0f
        if (rangeEnd > 1f) rangeEnd = 1f

        windowPaint.style = Paint.Style.FILL
        windowPaint.color = windowBgColor

        touchRipplePaint.style = Paint.Style.FILL
        touchRipplePaint.color = windowPaint.color

        chartPaint.style = Paint.Style.STROKE
        chartPaint.strokeWidth = chartLineWidth

        fadePaint.style = Paint.Style.FILL
        fadePaint.color = Color.argb((offWindowAlpha * 255).toInt(), 255, 255, 255)
        fadePaint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_OVER)

        touchSlop = ViewConfiguration.get(context).scaledTouchSlop
        dragSize = 1.5f * touchSlop
    }

    override fun onDraw(canvas: Canvas) {
        if (width == 0 || height == 0) return

        val width = width.toFloat()
        val height = height.toFloat()

        val rangeStartX = width * rangeStart
        val rangeEndX = width * rangeEnd

        // draw window
        val windowStart = rangeStartX + windowStrokeWidthLeftRight
        val windowEnd = rangeEndX - windowStrokeWidthLeftRight
        canvas.drawRect(0f, 0f, windowStart, height, windowPaint)
        canvas.drawRect(windowEnd, 0f, width, height, windowPaint)
        canvas.drawRect(windowStart, 0f, windowEnd, windowStrokeWidthTopBottom, windowPaint)
        canvas.drawRect(windowStart, height - windowStrokeWidthTopBottom, windowEnd, height, windowPaint)

        // draw Chart data
        graphs?.forEach { graph ->
            graph.draw(canvas, chartPaint)
        }

        // fade out-of-window range
        if (rangeStart > 0f) {
            canvas.drawRect(0f, 0f, rangeStartX, height, fadePaint)
        }
        if (rangeEnd < 1f) {
            canvas.drawRect(rangeEndX, 0f, width, height, fadePaint)
        }

        // draw touch ripple
        val touchRippleRadius = this.touchRippleRadius
        if (touchRippleRadius > 0f) {
            val cx = when (dragMode) {
                DRAG_RANGE_START -> rangeStartX + windowStrokeWidthLeftRight / 2f
                DRAG_RANGE_END -> rangeEndX - windowStrokeWidthLeftRight / 2f
                DRAG_RANGE -> rangeStartX + (rangeEndX - rangeStartX) / 2f
                else -> null
            }
            if (cx != null) {
                touchRipplePaint.alpha = (102 * touchRippleAlpha).toInt()
                canvas.drawCircle(cx, height / 2f, touchRippleRadius, touchRipplePaint)
            }
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        val action = event.action
        when (action) {
            MotionEvent.ACTION_DOWN -> {
                val rangeStartX = width * rangeStart
                val rangeEndX = width * rangeEnd

                downX = event.x
                lastTouchX = event.x
                inDragMode = false

                // determine drag mode depending on touch region
                dragMode = when (event.x) {
                    in (rangeStartX - dragSize)..(rangeStartX + dragSize + windowStrokeWidthLeftRight) -> DRAG_RANGE_START
                    in (rangeEndX - dragSize - windowStrokeWidthLeftRight)..(rangeEndX + dragSize) -> DRAG_RANGE_END
                    in (rangeStartX + dragSize + windowStrokeWidthLeftRight)..(rangeEndX - dragSize - windowStrokeWidthLeftRight) -> DRAG_RANGE
                    else -> DRAG_NONE
                }

                showTouchRippleAnimation()
            }
            MotionEvent.ACTION_MOVE -> {
                if (!inDragMode && Math.abs(event.x - downX) > touchSlop) {
                    inDragMode = true
                }
                if (inDragMode && dragMode != null && dragMode != DRAG_NONE) {
                    onMove(event.x, lastTouchX - event.x)
                    lastTouchX = event.x
                }
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                hideTouchRippleAnimation()
            }
        }
        return true
    }

    private fun onMove(eventX: Float, distanceX: Float) {
        val rangeStartX = width * rangeStart
        val rangeEndX = width * rangeEnd

        when (dragMode) {
            DRAG_RANGE_START -> {
                // move left edge of range window (0..rangeEnd)
                var newRangeStart = eventX / width
                if (newRangeStart < 0f) newRangeStart = 0f

                var newRangeStartX = width * newRangeStart
                if (newRangeStartX + dragSize + windowStrokeWidthLeftRight > rangeEndX - dragSize - windowStrokeWidthLeftRight) {
                    newRangeStartX = rangeEndX - 2f * (dragSize + windowStrokeWidthLeftRight)
                    newRangeStart = newRangeStartX / width
                }
                updateRanges(newRangeStart, rangeEnd)
            }
            DRAG_RANGE_END -> {
                // move right edge of range window (rangeStart..1)
                var newRangeEnd = eventX / width
                if (newRangeEnd > 1f) newRangeEnd = 1f

                var newRangeEndX = width * newRangeEnd
                if (newRangeEndX - dragSize - windowStrokeWidthLeftRight < rangeStartX + dragSize + windowStrokeWidthLeftRight) {
                    newRangeEndX = rangeStartX + 2f * (dragSize + windowStrokeWidthLeftRight)
                    newRangeEnd = newRangeEndX / width
                }
                updateRanges(rangeStart, newRangeEnd)
            }
            DRAG_RANGE -> {
                // move whole window
                var newRangeStart = (rangeStartX - distanceX) / width
                if (newRangeStart < 0f) newRangeStart = 0f

                var newRangeEnd = newRangeStart + (rangeEnd - rangeStart)
                if (newRangeEnd > 1f) {
                    newRangeEnd = 1f
                    newRangeStart = 1f - (rangeEnd - rangeStart)
                }
                updateRanges(newRangeStart, newRangeEnd)
            }
        }
    }

    private fun updateRanges(start: Float, end: Float) {
        if (rangeStart != start || rangeEnd != end) {
            rangeStart = start
            rangeEnd = end

            invalidate()

            onRangeChangeListener?.invoke(start, end)
        }
    }

    private fun showTouchRippleAnimation() {
        touchRippleAnimation?.cancel()
        touchRippleAnimation = ValueAnimator.ofFloat(0f, 1f).apply {
            duration = 200L
            interpolator = DecelerateInterpolator()
            addUpdateListener {
                touchRippleRadius = (it.animatedValue as Float) * (0.7f * height)
                touchRippleAlpha = it.animatedValue as Float
                postInvalidateOnAnimation()
            }
            start()
        }
    }

    private fun hideTouchRippleAnimation() {
        val currentValue = (touchRippleAnimation as? ValueAnimator)?.animatedValue as? Float ?: 1f
        touchRippleAnimation?.cancel()
        touchRippleAnimation = ValueAnimator.ofFloat(currentValue, 0f).apply {
            duration = 200L
            interpolator = AccelerateDecelerateInterpolator()
            addUpdateListener {
                touchRippleRadius = (it.animatedValue as Float) * (0.7f * height)
                touchRippleAlpha = it.animatedValue as Float
                postInvalidateOnAnimation()
            }
            start()
        }
    }

    // =============

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        uiChart?.let {
            enqueueUpdateData(it, width, height)
        }
    }

    fun attachToChartView(chartView: ChartView) {
        enqueueUpdateData(chartView.getUiChart(), width, height)

        chartView.setOnUiChartChangeListener { uiChart ->
            enqueueUpdateData(uiChart, width, height)
        }

        onRangeChangeListener = { start, end -> chartView.setRange(start, end) }

        chartView.setRange(rangeStart, rangeEnd)
    }

    private fun enqueueUpdateData(uiChart: UiChart?, viewWidth: Int, viewHeight: Int) {
        val updateGeneration = ++lastUpdateGeneration
        updateDataExecutor.submit {
            val newGraphs = updateData(uiChart, viewWidth, viewHeight)
            if (lastUpdateGeneration == updateGeneration) {
                post {
                    graphs = newGraphs
                    invalidate()
                }
            }
        }
    }

    private fun updateData(uiChart: UiChart?, viewWidth: Int, viewHeight: Int): List<Graph>? {
        if (uiChart == null) return null

        val scaleX = 1f * viewWidth.toFloat() / uiChart.width
        val scaleY = 0.85f * viewHeight / uiChart.height

        val scaledChartWidth = uiChart.width * scaleX
        val scaledChartHeight = uiChart.height * scaleY

        val xOffs = (width - scaledChartWidth) / 2f
        val yOffs = height - (height - scaledChartHeight) / 2f

        val matrix = Matrix()
        matrix.setTranslate(xOffs, yOffs)
        matrix.preScale(scaleX, -scaleY)

        return uiChart.graphs.map {
            it.transform(matrix)
        }
    }

}
